pkg/controller/clusterdeployment/clusterdeployment_controller.go

```diff
@@ -1956,101 +1952,59 @@ func (r *ReconcileClusterDeployment) deleteOldFailedProvisions(provs []*hivev1.C
-				return true
-			}
+// checkForFailedSync returns true if it finds that the ClusterSync has the Failed condition set
+func checkForFailedSync(clusterSync *hiveintv1alpha1.ClusterSync) bool {
+	for _, cond := range clusterSync.Status.Conditions {
+		if cond.Type == hiveintv1alpha1.ClusterSyncFailed {
+			return cond.Status == corev1.ConditionTrue
 		}
 	}
 	return false
 }
 
-// checkSyncSetConditionsForFailure returns true when the condition contains hivev1.ApplyFailureSyncCondition
-// and condition status is equal to true
-func checkSyncSetConditionsForFailure(conds []hivev1.SyncCondition) bool {
-	for _, c := range conds {
-		if c.Status != corev1.ConditionTrue {
-			continue
-		}
-		switch c.Type {
-		case hivev1.ApplyFailureSyncCondition, hivev1.DeletionFailedSyncCondition, hivev1.UnknownObjectSyncCondition:
-			return true
-		}
-	}
-	return false
-}
-
-// setSyncSetFailedCondition returns true when it sets or updates the hivev1.SyncSetFailedCondition
-func (r *ReconcileClusterDeployment) setSyncSetFailedCondition(cd *hivev1.ClusterDeployment, cdLog log.FieldLogger) (bool, error) {
-	// get all syncset instances for this cluster deployment
-	syncSetInstances, err := r.getAllSyncSetInstances(cd)
-	if err != nil {
-		cdLog.WithError(err).Error("Unable to list related syncset instances for cluster deployment")
-		return false, err
-	}
-
-	isFailedCondition := checkForFailedSyncSetInstance(syncSetInstances)
-
-	status := corev1.ConditionFalse
-	reason := "SyncSetApplySuccess"
-	message := "SyncSet apply is successful"
-	if isFailedCondition {
+// setSyncSetFailedCondition updates the hivev1.SyncSetFailedCondition
+func (r *ReconcileClusterDeployment) setSyncSetFailedCondition(cd *hivev1.ClusterDeployment, cdLog log.FieldLogger) error {
+	var (
+		status          corev1.ConditionStatus
+		reason, message string
+	)
+	clusterSync := &hiveintv1alpha1.ClusterSync{}
+	switch err := r.Get(context.Background(), types.NamespacedName{Namespace: cd.Namespace, Name: cd.Name}, clusterSync); {
+	case apierrors.IsNotFound(err):
+		cdLog.Info("ClusterSync has not yet been created")
+		status = corev1.ConditionFalse
+		reason = "MissingClusterSync"
+		message = "ClusterSync has not yet been created"
+	case err != nil:
+		cdLog.WithError(err).Log(controllerutils.LogLevel(err), "could not get ClusterSync")
+		return err
+	case checkForFailedSync(clusterSync):
 		status = corev1.ConditionTrue
 		reason = "SyncSetApplyFailure"
-		message = "One of the SyncSetInstance apply has failed"
+		message = "One of the SyncSet applies has failed"
+	default:
+		status = corev1.ConditionFalse
+		reason = "SyncSetApplySuccess"
+		message = "SyncSet apply is successful"
 	}
+
 	conds, changed := controllerutils.SetClusterDeploymentConditionWithChangeCheck(
 		cd.Status.Conditions,
 		hivev1.SyncSetFailedCondition,
 		status,
 		reason,
 		message,
-		controllerutils.UpdateConditionNever,
+		controllerutils.UpdateConditionIfReasonOrMessageChange,
```

Comment:
Hmm. That is not intentional. I should probably make the status True when the ClusterSync is missing. This will result in the condition being present usually, as the condition will b set to True for a short time until the syncsets are applied the first time. The other option is to not attempt set the condition at all until the ClusterSync is present. I don't expect a missing ClusterSync to be a common problem unless there are coding errors. I am leaning towards the former solution of setting the status to True.